### PR TITLE
[perf] cache type info for ParamEnv

### PR DESCRIPTION
**Pull request by lukas-code**
This is an attempt to mitigate some of the perf regressions in https://github.com/rust-lang/rust/pull/122553#issuecomment-2007563027, but seems worth to test and land separately, since it is mostly unrelated to that PR.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/list.rs [hunk 1]
```diff
@@ -193,7 +188,12 @@ impl<T> Deref for List<T> {
 impl<T> AsRef<[T]> for List<T> {
     #[inline(always)]
     fn as_ref(&self) -> &[T] {
-        unsafe { slice::from_raw_parts(self.data.as_ptr(), self.len) }
+        let data_ptr = ptr::addr_of!(self.skel.data).cast::<T>();
+        // SAFETY: `data_ptr` has the same provenance as `self` and can therefore
+        // access the `self.len` elements stored at `self.data`.
+        // Note that we specifically don't reborrow `&self.skel.data`, because that
+        // would give us a pointer with provenance over 0 bytes.
+        unsafe { slice::from_raw_parts(data_ptr, self.skel.len) }
```
Comment (lukas-code) [inline compiler/rustc_middle/src/ty/list.rs:196]: This was UB before.

File: compiler/rustc_middle/src/ty/list.rs [hunk 2]
```diff
@@ -193,7 +188,12 @@ impl<T> Deref for List<T> {
 impl<T> AsRef<[T]> for List<T> {
     #[inline(always)]
     fn as_ref(&self) -> &[T] {
-        unsafe { slice::from_raw_parts(self.data.as_ptr(), self.len) }
+        let data_ptr = ptr::addr_of!(self.skel.data).cast::<T>();
+        // SAFETY: `data_ptr` has the same provenance as `self` and can therefore
+        // access the `self.len` elements stored at `self.data`.
```
Comment (lukas-code) [inline compiler/rustc_middle/src/ty/list.rs:193]:
```suggestion
        // access the `self.skel.len` elements stored at `self.skel.data`.
```

**Comment by lukas-code** [comment]
uuh idk what bors is on about, the artifact looks like it got built? Let's try this:

:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8426883664/job/23076209908)
Build commit: be2a772ea7aa3f1114b45c8fb9dd2a2466a5c152 (`be2a772ea7aa3f1114b45c8fb9dd2a2466a5c152`)

**Comment by lukas-code** [comment]
well that worked i guess?

**Comment by lukas-code** [comment]
The regression in `token-stream-stress` looks like it's probably noise:

image

The regressions in `bitmaps` might be legit, but to me it seems like the improvements outweigh the regressions here.

r? @lcnr or compiler

File: compiler/rustc_middle/src/ty/list.rs [hunk 3]
```diff
@@ -210,23 +213,216 @@ unsafe impl<T: Sync> Sync for List<T> {}
...
-    const ALIGN: ptr::Alignment = {
+    const ALIGN: ptr::Alignment = align_of::<ListSkeleton<T>>();
+}
+
+/// A [`List`] that additionally stores type information inline to speed up
+/// [`TypeVisitableExt`](super::TypeVisitableExt) operations.
+#[repr(C)]
+pub struct ListWithCachedTypeInfo<T> {
+    skel: ListWithCachedTypeInfoSkeleton<T>,
+    opaque: OpaqueListContents,
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/list.rs:230]: why is that moved out of `ListSkeleton`?

**Review by lcnr — COMMENTED** [review]
r=me after explaining the changes in `ty/list.rs` to me :thinking:

File: compiler/rustc_middle/src/ty/list.rs (hunk 3 above)
Comment (lukas-code) [reply compiler/rustc_middle/src/ty/list.rs:230]: Because accessing a field with an extern type tail will always panic, because we don't know the align of the extern type and therefore the padding before the field. So basically if we just put `List` as the tail of `ListWithCachedTypeInfo`, then any attempt to access the list will panic.

File: compiler/rustc_middle/src/ty/list.rs [hunk 4]
```diff
@@ -210,23 +213,216 @@ unsafe impl<T: Sync> Sync for List<T> {}
...
+        (&**self).as_ref()
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for ListWithCachedTypeInfo<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+impl<'tcx, T: super::DebugWithInfcx<TyCtxt<'tcx>>> super::DebugWithInfcx<TyCtxt<'tcx>>
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/list.rs:371]: nit: please instead import `DebugWithInfcx`

File: compiler/rustc_middle/src/ty/list.rs (hunk 3 above)
Comment (lukas-code) [reply compiler/rustc_middle/src/ty/list.rs:230]: Hmmm apparently there is a special case if the field with an extern type tail is the only field. Which i guess makes sense since there can't be any padding in that case. If you change the outer type to have some extra data, your example will panic: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=178bcc676aa494d0452be7b3baeb5d83

> I don't think we should ever copy `ListSkeleton` to the stack

I agree. If my code does that it's not intended.

Comment (lukas-code) [reply compiler/rustc_middle/src/ty/list.rs:245]: I've added some comments to the `*Skeleton` types to hopefully clear this up a bit.

Relevant diff: https://github.com/rust-lang/rust/compare/c4f1f7aa586451154c669711abc9c4d9ad9b1075..fcc477fbd0871d8fab045bd7e9524172ab10e51c

**Review by lcnr — COMMENTED** [review]
:thinking: the panic with extern types is annoying, but I understand the issue now.

Looking at this again, we now duplicate the unsafe code for `List` which is somewhat intricate and I am a bit worried that it ends up getting out of sync. Could we instead have a single type:
```rust
struct RawList<H, T> {
    header: H,
    length: usize,
    data: [T; 0],
    opaque: OpaqueListConten,
}
```
This should avoid both the "extern type alignment" issue and aloows us to remove the code duplication.

Sorry for not thinking of that during the first review round 😅

**Comment by nnethercote** [comment]
I stumbled across this today while looking at `List`. I see `RawList` has been added without much explanation in the comments. It seems to allow an arbitrary header to be included before the list length. How/why is this a speed win in some cases? It seems very strange to augment a generic container type with an arbitrary value of a different type.

**Comment by lukas-code** [comment]
There are some functions in the trait solver that use the [`TypeVisitableExt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.TypeVisitableExt.html) methods on a [`ParamEnv`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.ParamEnv.html) to find out some property about the where clauses that are assumed to be true during trait solving.

For example here we disable caching here if the where clauses contain inference variables (which can happen due to [canonicalization](https://rust-lang.github.io/chalk/book/canonical_queries/canonicalization.html)).
https://github.com/rust-lang/rust/blob/5f8c17dcc04a2981268df89874203e9bfea50597/compiler/rustc_trait_selection/src/traits/select/mod.rs#L1508

Or here we find out whether inference has happened on the where clauses by resolving the inference variables. This has a [shortcut](https://github.com/rust-lang/rust/blob/5f8c17dcc04a2981268df89874203e9bfea50597/compiler/rustc_infer/src/infer/mod.rs#L1374-L1376) if there are no inference variables.
https://github.com/rust-lang/rust/blob/5f8c17dcc04a2981268df89874203e9bfea50597/compiler/rustc_trait_selection/src/traits/select/mod.rs#L574

---

Before this PR, the `ParamEnv` was represented roughly like this:
```text
               ------------------------------------------------------------
Clause         | type flags | outer binder | stable hash | predicate kind |
               ------------------------------------------------------------
               ^
               |
               +------------+
                            |
               -------------------------------------
List<Clause>   | length | clause 1 | clause 2 | ... 
               -------------------------------------
                    ^
                    |
                    |
               --------------------------
ParamEnv       | pointer   | reveal bit |
               --------------------------
```

So to find out the type flags or outer binder of the whole `ParamEnv` we'd have to iterate over all of the clauses. For some crates, in particular `diesel`, the `ParamEnv`s can get rather large (order of magnitude 100..1000) and iterating over them several times for every trait that we have to check adds up.

The main thing this PR does is add `ListWithCachedTypeInfo`, which stores the type flags and outer binder directly in front of the `List`, and use that new type in `ParamEnv`:
```text
               ------------------------------------------------------------
Clause         | type flags | outer binder | stable hash | predicate kind |
               ------------------------------------------------------------
               ^
               |
               +--------------------------------------------------------+
                                                                        |
                                 -----------------------------------------------------------------
ListWithCachedTypeInfo<Clause>   | type flags | outer binder | length | clause 1 | clause 2 | ... 
                                 -----------------------------------------------------------------
                                  ^
                                  |
                    +-------------+
                    |
               --------------------------
ParamEnv       | pointer   | reveal bit |
               --------------------------
```

This way, we can compute e.g. `param_env.has_infer()` in **O(1)** rather than **O(n)**.

`RawList` only exists to share code between `List<T>` (= `RawList<(), T>`) and `ListWithCachedTypeInfo<T>` (= `RawList<TypeInfo, T>`).

**Comment by nnethercote** [comment]
Thanks for the explanation. In this bit:
```
                                 -----------------------------------------------------------------
ListWithCachedTypeInfo<Clause>   | type flags | outer binder | length | clause 1 | clause 2 | ... 
                                 -----------------------------------------------------------------
```
are the `type flags`/`outer binder` computed as a combination of the same-named fields from all the clauses?

**Comment by lukas-code** [comment]
> are the `type flags`/`outer binder` computed as a combination of the same-named fields from all the clauses?

Exactly. The flags of the list is the bit-or of all the flags from the clauses and the outer binder of the list is the maximum of the outer binders of the clauses. 

They are computed when the `ListWithCachedTypeInfo<Clauses>` is constructed here:
https://github.com/rust-lang/rust/blob/1189851eebdf6b69b164bbb7bf9df91188ae03aa/compiler/rustc_middle/src/ty/context.rs#L406-L412
https://github.com/rust-lang/rust/blob/1189851eebdf6b69b164bbb7bf9df91188ae03aa/compiler/rustc_middle/src/ty/flags.rs#L38-L45
